rtl/eclrtl/eclregex.cpp

```diff
@@ -393,55 +435,176 @@ class CCompiledStrRegExpr : implements ICompiledStrRegExpr
         }
         else
         {
             // No match found; return the original string
             out = (char *)rtlMalloc(sourceSize);
             memcpy_iflen(out, str, sourceSize);
             outlen = slen;
-            pcre2_match_data_free_8(matchData);
+        }
+    }
+
+    // This method supports "fixed length UTF-8" even though that isn't really a thing;
+    // it's here more for completeness, in case we ever implement some version of it
+    void replaceFixed(size32_t tlen, char * tgt, size32_t slen, char const * str, size32_t rlen, char const * replace) const
+    {
+        if (tlen == 0)
+            return;
+
+        PCRE2MatchData8 matchData = pcre2_match_data_create_from_pattern_8(compiledRegex.get(), pcre2GeneralContext8);
+
+        // This method is often called through an ECL interface and the provided lengths
+        // (slen and rlen) are in code points (characters), not bytes; we need to convert these to a
+        // byte count for PCRE2
+        size32_t sourceSize = (isUTF8Enabled ? rtlUtf8Size(slen, str) : slen);
+        size32_t replaceSize = (isUTF8Enabled ? rtlUtf8Size(rlen, replace) : rlen);
+
+        // Execute an explicit match first to see if we match at all; if we do, matchData will be populated
+        // with data that can be used by pcre2_substitute to bypass some work
+        int numMatches = pcre2_match_8(compiledRegex.get(), (PCRE2_SPTR8)str, sourceSize, 0, 0, matchData, pcre2MatchContext8);
+
+        if (numMatches < 0 && numMatches != PCRE2_ERROR_NOMATCH)
+        {
+            // Treat everything other than PCRE2_ERROR_NOMATCH as an error
+            failWithPCRE2Error(numMatches, "Error in regex replace: ");
+        }
+
+        if (numMatches > 0)
+        {
+            uint32_t replaceOptions = PCRE2_SUBSTITUTE_MATCHED|PCRE2_SUBSTITUTE_GLOBAL|PCRE2_SUBSTITUTE_EXTENDED;
+            PCRE2_SIZE pcreSize = 0;
+
+            // Call substitute once to get the size of the output and see if it will fit within fixedOutLen;
+            // if it does then we can substitute within the given buffer and then pad with spaces, if not then
+            // we have to allocate memory, substitute into that memory, then copy into the given buffer;
+            // note that pcreSize will include space for a terminating null character even though we don't want it
+            int replaceResult = pcre2_substitute_8(compiledRegex.get(), (PCRE2_SPTR8)str, sourceSize, 0, replaceOptions|PCRE2_SUBSTITUTE_OVERFLOW_LENGTH, matchData, pcre2MatchContext8, (PCRE2_SPTR8)replace, replaceSize, nullptr, &pcreSize);
+
+            if (replaceResult < 0 && replaceResult != PCRE2_ERROR_NOMEMORY)
+            {
+                // PCRE2_ERROR_NOMEMORY is a normal result when we're just asking for the size of the output;
+                // everything else is an error
+                failWithPCRE2Error(replaceResult, "Error in regex replace: ");
+            }
+
+            if (pcreSize > 0)
```

Comment:
for no output string I think pcreSize will be 1.  What does pcreSize==0 represent?  Should there be any code in the else?

Reply:
@ghalliday back to you.  Thanks!